projects/control-service/projects/pipelines_control_service/src/main/java/com/vmware/taurus/service/deploy/DeploymentService.java

```diff
@@ -141,6 +144,8 @@ public void updateDeployment(
     try {
       log.info("Starting deployment of job {}", jobDeployment.getDataJobName());
       deploymentProgress.started(dataJob.getJobConfig(), jobDeployment);
+      validatePythonVersionIsSupported(jobDeployment.getPythonVersion());
```

Comment:
The validation cannot be done at the `updateDeployment` level since it is asynchronous and in case of an incorrect python version, it won't return an error to the client.

Reply:
+1 validation should happen on the synchronous path so that API request fail immediately. 

On the async path. If it happens that the configured python version is not part of the supported version (it could happen if administrators change the supported python versions -e.g let's say admins remove for 3.7 after end of life support) , but user has configured 3.7 

What do we do? 
My suggest was 

A) To log a warning so the administrator is aware (as much as it's expected that admins look at warnings ...) 
B) Do not fail the deployment for certain. 
  B.1) It's possible to continue using the python version (but as it's not supported there's risk the image would not exist and generally I do not favour that option 
  B.2) Find closest supported version and use it.  
C) After B) it might be a good idea to send a notification to user as well so they are aware their job was "forcefully" upgraded  

I know we discussed it briefly during the vep discussion but I do not recall if we reached any consuliton on that. 

It is perfectly fine if you decide to tacke that part of the problem in separate PR.

Reply:
Alternatively we can add the validate not just on patch but also on any deploy operation , 
So the next time user tries to deploy they'd get an error that tells them to update the version. 

I prefer that we handle this for the user. Generally from UX point of view, reducing things users need to think about is always preferable. But it's an option.

Reply:
Good points, moved the validation to the Deployment Controller.

> B.2) Find closest supported version and use it.
I am not a big fan of such approach. How do you define "closest supported version"? If we have ["3.7", "3.8", "3.9"] and the user passes "3.10", then sure the closest would be 3.9. However, if we have ["3.8", "3.10", "3.12"] and the user passes "3.9", what would be the closest - 3.8 or 3.10?
Also, there is no limitation on the format, so the admins can set ["3.8.12", "3.9rc2", "3.10"] and these would still be valid choices. The parsing logic to find the closest version would be a nightmare and prone to corner case issues.

Reply:
You will not be parsing it yourself. You will be using a library like https://github.com/vdurmont/semver4j . 
The logic should not be complicated. sort them all and find the next closest.